### PR TITLE
WIP: Init command

### PR DESCRIPTION
**Pull request by ilyavolodin**
This is a WIP request for init command. I'm not really sure that eslint.js is the right place to put it in, maybe it should go into it's own file? Also please check the questions and see if we would like to add any others?
Eslint.js is currently not unittested. So I didn't create any unit tests for init either for now (I'm not really sure they are necessary either).

### CODE REVIEW

**Comment by gyandeeps** [comment]
My recommendation would be to put it into its ow file so that it can be easily tested.

File: bin/eslint.js [hunk 1]
```diff
@@ -15,6 +19,37 @@ if (useStdIn) {
...
+        { type: "list", name: "quotes", message: "What quotes do you use for strings?", default: "double", choices: [{ name: "Double", value: "double" }, { name: "Single", value: "single" }] },
+        { type: "list", name: "linebreak", message: "What line endings do you use?", default: "unix", choices: [{ name: "Unix", value: "unix" }, { name: "Windows", value: "windows" }]},
+        { type: "confirm", name: "semi", message: "Do you require semicolons?", default: true},
+        { type: "confirm", name: "es6", message: "Are you using ECMAScript 6 features?", default: false},
+        { type: "list", name: "env", message: "Where will your code run?", default: "browser", choices: [{ name: "Node", value: "node" }, { name: "Browser", value: "browser" }]},
+        { type: "confirm", name: "jsx", message: "Will you use JSX?", default: false},
+        { type: "confirm", name: "react", message: "Will you use React", default: false, when: function(answers) { return answers.jsx; }},
+        { type: "list", name: "format", message: "What format do you want your config file to be in?", default: "JSON", choices: ["JSON", "YAML"]}
```
Comment (gyandeeps) [inline bin/eslint.js]: I think we should convert these single lines into multi-line as it would increase the readability of the code.

**Comment by nzakas** [comment]
Yeah, I think I'd create a new object (ConfigInitializer?) and then just reference that from eslint.js.

File: bin/eslint.js [hunk 2]
```diff
@@ -15,6 +19,37 @@ if (useStdIn) {
...
         }
     }));
+} else if (init) {
+    inquirer.prompt([
+        { type: "list", name: "indent", message: "What style of indentation do you use?", default: "tabs", choices: [{ name: "Tabs", value: "tabs" }, { name: "Spaces", value: 4}] },
+        { type: "list", name: "quotes", message: "What quotes do you use for strings?", default: "double", choices: [{ name: "Double", value: "double" }, { name: "Single", value: "single" }] },
+        { type: "list", name: "linebreak", message: "What line endings do you use?", default: "unix", choices: [{ name: "Unix", value: "unix" }, { name: "Windows", value: "windows" }]},
+        { type: "confirm", name: "semi", message: "Do you require semicolons?", default: true},
+        { type: "confirm", name: "es6", message: "Are you using ECMAScript 6 features?", default: false},
+        { type: "list", name: "env", message: "Where will your code run?", default: "browser", choices: [{ name: "Node", value: "node" }, { name: "Browser", value: "browser" }]},
```
Comment (nzakas) [inline bin/eslint.js]: We need to be able to select both node and browser.

Comment (ilyavolodin) [inline bin/eslint.js]: Ahh.. right, forgot that they are not mutually exclusive. Good thing inquirer has checkboxes prompt.

**Comment by ilyavolodin** [comment]
Updated pull request. I don't really like having to send 3 function pointers into a function, but I don't really see a way around it as inquirer is async.

File: bin/configInitializer.js [hunk 3]
```diff
@@ -0,0 +1,131 @@
+"use strict";
+
```
Comment (nzakas) [inline bin/configInitializer.js]: This file should be in /lib and have the same header comments as all other files.

File: bin/configInitializer.js [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+/* istanbul ignore next: hard to test fs function */
+/**
+ * Create .eslintrc file in the current working directory
+ * @param {object} config object that contains user's answers
+ * @param {bool} isJson should config file be json or yaml
+ * @param {function} callback callback function when finished writing .eslintrc file
+ * @returns {void}
+ */
+function writeFile(config, isJson, callback) {
+    fs.writeFile("./.eslintrc", isJson ? JSON.stringify(config, null, 4) : yaml.safeDump(config), callback);
```
Comment (nzakas) [inline bin/configInitializer.js]: Why not just writeFileSync? Eliminate one callback.

File: bin/configInitializer.js [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+        {
+            type: "list",
+            name: "format",
+            message: "What format do you want your config file to be in?",
+            default: "JSON",
+            choices: ["JSON", "YAML"]
+        }
+    ], function(answers) {
+        processConfigFn(answers, createConfigFileFn, callback);
+    });
```
Comment (nzakas) [inline bin/configInitializer.js]: I think this callback is doing too much. This function should just gather answers and call a single callback. It should be up to the caller to decide what to do.

File: bin/configInitializer.js [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+var init = {
+    /**
+     * process user's answers and create config object
+     * @param {object} answers answers received from inquirer
+     * @param {function} createConfigFile function to write config file to disk
+     * @param {function} callback function to be called when file is written
+     * @returns {void}
+     */
+    processAnswers: function(answers, createConfigFile, callback) {
+        var config = {rules: {}, env: {}};
```
Comment (nzakas) [inline bin/configInitializer.js]: I'd just have this function return an object with the processed values. Much easier to test.

File: bin/configInitializer.js [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+            config.env[env] = true;
+        });
+        if (answers.jsx) {
+            config.ecmaFeatures = {jsx: true};
+            if (answers.react) {
+                config.plugins = ["react"];
+            }
+        }
+        createConfigFile(config, answers.format === "JSON", callback);
+    },
```
Comment (nzakas) [inline bin/configInitializer.js]: Also need to npm install eslint-plugin-react

**Comment by nzakas** [comment]
I think this can be simplified by better breaking up the various steps. Inquirer is a sync, but that doesn't mean everything has to be.

**Comment by ilyavolodin** [comment]
Sorry, last commit should've never been pushed. Guess I shouldn't be pushing code at midnight. I think it's better now. The only thing I still don't like is exposing `processAnswers` publicly. But that's the only way I can test it:-(

File: lib/configInitializer.js [hunk 8]
```diff
@@ -0,0 +1,140 @@
+/**
```
Comment (nzakas) [inline lib/configInitializer.js]: filename: config-initializer.js to match the others

**Comment by nzakas** [comment]
Looks cleaner now. We still need to add in the call to `npm i eslint-plugin-react --save-dev` if they are using React.

**Comment by ilyavolodin** [comment]
@nzakas Sorry, forgot about that. Added it and changed file names.

File: lib/config-initializer.js [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+            type: "checkbox",
+            name: "env",
+            message: "Where will your code run?",
+            default: ["browser"],
+            choices: [{name: "Node", value: "node"}, {name: "Browser", value: "browser"}]
+        },
+        {
+            type: "confirm",
+            name: "jsx",
+            message: "Will you use JSX?",
```
Comment (nzakas) [inline lib/config-initializer.js]: I think "Do you use JSX?" to be consistent with other questions

File: lib/config-initializer.js [hunk 10]
```diff
@@ -0,0 +1,146 @@
...
+            default: false
+        },
+        {
+            type: "confirm",
+            name: "react",
+            message: "Will you use React",
```
Comment (nzakas) [inline lib/config-initializer.js]: "Do you use React?"

**Comment by nzakas** [comment]
Looks good, just a couple wording issues

**Comment by ilyavolodin** [comment]
Done

**Comment by nzakas** [comment]
Sweet, this is really cool.

**Pull request merged**